### PR TITLE
ISSUE-1272 Team mailboxes over IMAP

### PR DESCRIPTION
**Pull request by vttranlina**
resolve https://github.com/linagora/tmail-backend/issues/1272

### CODE REVIEW

File: tmail-backend/imap-extensions/src/test/java/imap/linagora/tmail/imap/TMailMailboxNameConverterTest.java [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+import org.apache.james.mailbox.MailboxSessionUtil;
+import org.apache.james.mailbox.model.MailboxConstants;
+import org.apache.james.mailbox.model.MailboxPath;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+import com.linagora.tmail.imap.TMailMailboxNameConverter;
+
+public class TMailMailboxNameConverterTest {
```
Comment (chibenwa) [inline tmail-backend/imap-extensions/src/test/java/imap/linagora/tmail/imap/TMailMailboxNameConverterTest.java:17]: Please ensure 100% code coverage

File: tmail-backend/integration-tests/imap/memory-imap-integration-tests/src/test/java/com/linagora/tmail/imap/IMAPTeamMailboxIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            .create(MARKETING_TEAM_MAILBOX)
+            .create(SALE_TEAM_MAILBOX)
+            .addMember(MARKETING_TEAM_MAILBOX, MINISTER)
+            .addMember(SALE_TEAM_MAILBOX, MINISTER);
+
+        imapPort = server.getProbe(ImapGuiceProbe.class).getImapPort();
+        System.out.println("ImapPort: " + imapPort);
+    }
+
+//    @Test
```
Comment (chibenwa) [inline tmail-backend/integration-tests/imap/memory-imap-integration-tests/src/test/java/com/linagora/tmail/imap/IMAPTeamMailboxIntegrationTest.java:101]: Why?

**Comment by chibenwa** [comment]
And screenshots in thunderbird please ;-)

File: tmail-backend/integration-tests/imap/memory-imap-integration-tests/src/test/java/com/linagora/tmail/imap/IMAPTeamMailboxIntegrationTest.java (hunk 2 above)
Comment (vttranlina) [reply tmail-backend/integration-tests/imap/memory-imap-integration-tests/src/test/java/com/linagora/tmail/imap/IMAPTeamMailboxIntegrationTest.java]: local draft test
-> removed

**Comment by vttranlina** [comment]
> And screenshots in thunderbird please ;-)

image

**Comment by vttranlina** [comment]
> And screenshots in thunderbird please ;-)

image

**Review by quantranhong1999 — APPROVED** [review]
Good job.

File: tmail-backend/imap-extensions/src/test/java/imap/linagora/tmail/imap/TMailMailboxNameConverterTest.java [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+                .isEqualTo("#Shared.marketing.abc");
+        }
+    }
+    /*
+        Copied from org.apache.james.imap.main.PathConverterTest
+        ---END---
+     */
+
+    @Test
+    void buildFullPathShouldAcceptTeamMailboxName() {
```
Comment (chibenwa) [inline tmail-backend/imap-extensions/src/test/java/imap/linagora/tmail/imap/TMailMailboxNameConverterTest.java:287]: We do not test `mailboxQuery` and `mailboxName` for team mailboxes?

File: tmail-backend/imap-extensions/src/test/java/imap/linagora/tmail/imap/TMailMailboxNameConverterTest.java [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+        }
+
+        @Test
+        void mailboxNameShouldReturnFQDNWhenNotRelativeAndSharedMailbox() {
+            assertThat(pathConverter.mailboxName(!RELATIVE, new MailboxPath("#Shared", Username.of("marketing@apache.org"), "abc"), mailboxSession))
+                .isEqualTo("#Shared.marketing.abc");
+        }
+    }
+    /*
+        Copied from org.apache.james.imap.main.PathConverterTest
```
Comment (chibenwa) [inline tmail-backend/imap-extensions/src/test/java/imap/linagora/tmail/imap/TMailMailboxNameConverterTest.java:282]: This copy is a shame.

How about writing on James side a `PathConverterBasicContract` that PathConverterTest and `TMailMailboxNameConverterTest` could reuse?

File: tmail-backend/imap-extensions/src/main/java/com/linagora/tmail/imap/TMailMailboxNameConverter.java [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+import com.google.common.collect.Iterables;
+import com.google.common.escape.Escaper;
+import com.google.common.escape.Escapers;
+import com.linagora.tmail.team.TeamMailbox;
+import com.linagora.tmail.team.TeamMailboxNameSpace;
+
+/**
+ * Fork from org.apache.james.imap.main.PathConverter.Default
+ */
+public class TMailMailboxNameConverter implements PathConverter {
```
Comment (chibenwa) [inline tmail-backend/imap-extensions/src/main/java/com/linagora/tmail/imap/TMailMailboxNameConverter.java:30]: I am not OK with the level of duplication of this class.

I would rather prefer using `DefaultPathConverter` as a delegate when we are without namespace, `#private` and `#user` while writing custom code to manage `#team-mailboxes`.

This copy means we can have a dangerous delta with James code, meaning any bug fixed upstream won't be fixed here. We shall duplicate code to the minimum.

File: tmail-backend/integration-tests/imap/memory-imap-integration-tests/src/test/java/com/linagora/tmail/imap/IMAPTeamMailboxIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+    @Test
+    void namespaceShouldReturnTeamMailboxNameSpace() throws Exception {
+        assertThat(testIMAPClient.connect(IMAP_HOST, imapPort)
+            .login(MINISTER, MINISTER_PASSWORD)
+            .sendCommand("NAMESPACE"))
+            .contains("NAMESPACE ((\"\" \".\")) ((\"#user.\" \".\")) ((\"#TeamMailbox.\" \".\"))");
+    }
+
+    @Test
+    void listShouldReturnAllTeamMailboxWhenTeamMailboxNamespace() throws Exception {
```
Comment (chibenwa) [inline tmail-backend/integration-tests/imap/memory-imap-integration-tests/src/test/java/com/linagora/tmail/imap/IMAPTeamMailboxIntegrationTest.java:121]: Please write all variations of `LIST` similar of tests written for `#user`:

```
a0 LIST "" "*"
a1 LIST "#TeamMailbox" "*"
a2 LIST "#TeamMailbox" "marketting.*"
a3 LIST "#TeamMailbox.marbetting" "*"
a4 LIST "" "#TeamMailbox.*"
a5 LIST "" "#TeamMailbox.marketting.*"
```

Better safe than sorry...

**Comment by vttranlina** [comment]
Requirement: 
 james-code 
- https://github.com/apache/james-project/pull/2512

**Review by chibenwa — COMMENTED** [review]
Looks like good quality work.

In the end @vttranlina did using composition, rather than inheritance, lead to more boiler plate code being written?

**Review by chibenwa — APPROVED** [review]

**Comment by vttranlina** [comment]
Not as much boilerplate code as I initially assumed.

**Review by quantranhong1999 — APPROVED** [review]

**Comment by quantranhong1999** [comment]
`[ERROR] /home/jenkins/build/workspace/Tmail_build_PR-1311/tmail-backend/imap-extensions/src/main/java/com/linagora/tmail/imap/TMailPathConverter.java:[56,52] incompatible types: java.lang.String cannot be converted to java.util.Optional<java.lang.String>`

**Pull request merged**